File: legion/packager/rest/legion/packager/rest/extensions/ecr.py
#
#    Copyright 2019 EPAM Systems
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
#

import base64
import logging
import re
import typing
from typing import NamedTuple

from boto3 import session
from botocore.exceptions import ClientError
from legion.sdk.models import ConnectionSpec

# Find the regular expression here https://github.com/awslabs/amazon-ecr-credential-helper/blob/c689f646af7fc73f3e4cea0a8dd4671eeeb6884e/ecr-login/api/client.go#L34  # pylint: disable=C0301
ECR_URL = re.compile(
    r'(^[a-zA-Z0-9][a-zA-Z0-9-_]*)\.dkr\.ecr(\-fips)?\.([a-zA-Z0-9][a-zA-Z0-9-_]*)\.amazonaws\.com(\.cn)?\/?(.*):.*$')
ECR_CONNECTION_TYPE = 'ecr'


class ECRRegistry(NamedTuple):
    id: str
    repository_name: str


def get_ecr_credentials(conn_spec: ConnectionSpec, image_docker_url: str) -> typing.Tuple[str, str]:
    """
    The function generates docker credentials for the ECR registry.
    It creates a repository for an image if there is not it.
    :param conn_spec: Connection Spec
    :param image_docker_url: full docker image URI
    :return:
    """
    if conn_spec.type != ECR_CONNECTION_TYPE:
        raise ValueError(f'Unexpected connection type: {conn_spec.type}')

    ecr_client = session.Session(
        aws_access_key_id=conn_spec.key_id,
        aws_secret_access_key=conn_spec.key_secret,
        region_name=conn_spec.region).client('ecr')

    ecr_registry = extract_repository_from_ecr_url(image_docker_url)

    try:
        ecr_client.create_repository(repositoryName=ecr_registry.repository_name,
                                     imageTagMutability='MUTABLE')

        logging.info(f'Repository {ecr_registry.repository_name} was created')
    except ClientError as e:
        if e.response['Error']['Code'] == 'RepositoryAlreadyExistsException':
            logging.info(f'Repository {ecr_registry.repository_name} has already existed')
        else:
            raise e

    response = ecr_client.get_authorization_token(registryIds=[ecr_registry.id])
    token = response["authorizationData"][0]["authorizationToken"]

    return base64.b64decode(token).decode().split(':')


def extract_repository_from_ecr_url(image_url: str) -> ECRRegistry:
    """
    Retrieve repository information from ecr URL
    :param image_url: full docker image URI
    :return: repository information
    """
    match = ECR_URL.search(image_url)
    if match is None:
        raise ValueError(f'{image_url} is invalid ecr image')

    groups = match.groups()

    if len(groups) != 5:
        raise ValueError(f'{image_url} is not a valid repository URI for Amazon Elastic Container Registry.')

    return ECRRegistry(id=groups[0], repository_name=groups[4])

File: legion/operator/pkg/utils/aws/ecr.go
//
//    Copyright 2019 EPAM Systems
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
//

package aws

import (
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/endpoints"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/awslabs/amazon-ecr-credential-helper/ecr-login/api"
	legionv1alpha1 "github.com/legion-platform/legion/legion/operator/pkg/apis/legion/v1alpha1"
	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
)

var log = logf.Log.WithName("legion-ecr")

// Retrieves the docker credentials from ecr registry
func ExtractEcrCreds(connSpec legionv1alpha1.ConnectionSpec) (user string, password string, err error) {
	registry, err := api.ExtractRegistry(connSpec.URI)
	if err != nil {
		log.Error(err, "Error parsing the serverURL", "serverURL", connSpec.URI)
		return
	}

	var client api.Client
	if registry.FIPS {
		client, err = newClientWithFipsEndpoint(connSpec)
		if err != nil {
			log.Error(err, "Error resolving FIPS endpoint")

			return "", "", err
		}
	} else {
		client, err = newClientFromRegion(connSpec)
		if err != nil {
			log.Error(err, "Error creating client")

			return "", "", err
		}
	}

	auth, err := client.GetCredentials(connSpec.URI)
	if err != nil {
		log.Error(err, "Error retrieving credentials")

		return "", "", err
	}
	return auth.Username, auth.Password, nil
}

func newSession(connSpec legionv1alpha1.ConnectionSpec) (*session.Session, error) {
	return session.NewSession(&aws.Config{
		Region:      aws.String(connSpec.Region),
		Credentials: credentials.NewStaticCredentials(connSpec.KeyID, connSpec.KeySecret, ""),
	})
}

// NewClientFromRegion uses the region to create the client
func newClientFromRegion(connSpec legionv1alpha1.ConnectionSpec) (api.Client, error) {
	awsSession, err := newSession(connSpec)
	if err != nil {
		log.Error(err, "Session creation")

		return nil, err
	}

	awsConfig := &aws.Config{Region: aws.String(connSpec.Region)}
	return api.DefaultClientFactory{}.NewClientWithOptions(api.Options{
		Session: awsSession,
		Config:  awsConfig,
	}), nil
}

// NewClientWithFipsEndpoint overrides the default ECR service endpoint in a given region to use the FIPS endpoint
func newClientWithFipsEndpoint(connSpec legionv1alpha1.ConnectionSpec) (api.Client, error) {
	awsSession, err := newSession(connSpec)
	if err != nil {
		log.Error(err, "Session creation")

		return nil, err
	}

	endpoint, err := getServiceEndpoint("ecr-fips", connSpec.Region)
	if err != nil {
		return nil, err
	}

	awsConfig := awsSession.Config.WithEndpoint(endpoint).WithRegion(connSpec.Region)
	return api.DefaultClientFactory{}.NewClientWithOptions(api.Options{
		Session: awsSession,
		Config:  awsConfig,
	}), nil
}

func getServiceEndpoint(service, region string) (string, error) {
	resolver := endpoints.DefaultResolver()
	endpoint, err := resolver.EndpointFor(service, region, func(opts *endpoints.Options) {
		opts.ResolveUnknownService = true
	})

	if err != nil {
		return "", err
	}

	return endpoint.URL, err
}
